AppCenterAuth/AppCenterAuth/MSAuth.m

```diff
@@ -180,14 +181,36 @@ + (void)signOut {
 
 - (void)signIn {
   if ([[MS_Reachability reachabilityForInternetConnection] currentReachabilityStatus] == NotReachable) {
-    [self completeSignInWithErrorCode:MSACAuthErrorSignInWhenNoConnection andMessage:@"User sign-in failed. Internet connection is down."];
+    [self ifSignInIsInProgressCompleteWithErrorCode:MSACAuthErrorSignInWhenNoConnection
+                                         andMessage:@"User sign-in failed. Network is unavailable."
+                            isDownloadConfigFailure:NO];
     return;
   }
   if (self.clientApplication == nil || self.authConfig == nil) {
-    [self completeSignInWithErrorCode:MSACAuthErrorSignInBackgroundOrNotConfigured
-                           andMessage:@"signIn is called while it's not configured or not in the foreground."];
+    if (self.signInShouldWaitForConfig) {
+      MSLogDebug([MSAppCenter logTag], @"Downloading configuration in process. Waiting for it before sign-in.");
+    } else {
+      [self ifSignInIsInProgressCompleteWithErrorCode:MSACAuthErrorSignInBackgroundOrNotConfigured
+                                           andMessage:@"signIn is called while it's not configured or not in the foreground."
+                              isDownloadConfigFailure:NO];
+    }
     return;
   }
+  [self continueSignInAndStopWaitingForConfig:NO];
+}
+
+- (void)continueSignInAndStopWaitingForConfig:(BOOL)stopWaitingForConfig {
+  @synchronized(self) {
```

Comment:
Then the sign-in API should be synchronized too since it's relying on `signInShouldWaitForConfig`

Reply:
@clpolet outdated